nexus/src/app/background/instance_watcher.rs

```diff
@@ -0,0 +1,635 @@
+    /// removed from the database between when we listed instances and when the
+    /// check completed.
+    InstanceNotFound,
+    /// Something went wrong while updating the state of the instance in the
+    /// database.
+    UpdateFailed,
+}
+
+impl fmt::Display for CheckFailure {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::SledAgentUnreachable => f.write_str("unreachable"),
+            Self::SledAgentResponse(status) => {
+                write!(f, "{status}")
+            }
+            Self::NoSuchInstance => f.write_str("no_such_instance"),
+        }
+    }
+}
+
+impl fmt::Display for CheckError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::ClientHttpError(status) => write!(f, "{status}"),
+            Self::ClientError => f.write_str("client_error"),
+            Self::InstanceNotFound => f.write_str("instance_not_found"),
+            Self::UpdateFailed => f.write_str("update_failed"),
+        }
+    }
+}
+
+type ClientError = sled_agent_client::Error<sled_agent_client::types::Error>;
+
+impl BackgroundTask for InstanceWatcher {
+    fn activate<'a>(
+        &'a mut self,
+        opctx: &'a OpContext,
+    ) -> BoxFuture<'a, serde_json::Value> {
+        async {
+            let mut tasks = tokio::task::JoinSet::new();
+            let mut paginator = Paginator::new(MAX_SLED_AGENTS);
+            while let Some(p) = paginator.next() {
+                let maybe_batch = self
+                    .datastore
+                    .sled_instance_list_by_sled_agent(
+                        opctx,
+                        &p.current_pagparams(),
+                    )
+                    .await;
+                let batch = match maybe_batch {
+                    Ok(batch) => batch,
+                    Err(e) => {
+                        slog::warn!(
+                            opctx.log,
+                            "sled instances by sled agent query failed: {e}"
+                        );
+                        break;
+                    }
+                };
+                paginator = p.found_batch(&batch, &|(sled, _)| sled.id());
+                let mut batch = batch.into_iter();
+
+                if let Some((mut curr_sled, sled_instance)) = batch.next() {
+                    let mk_client = |sled: &Sled| {
+                        nexus_networking::sled_client_from_address(
+                            sled.id(),
+                            sled.address(),
+                            &opctx.log,
+                        )
+                    };
+
+                    let mut client = mk_client(&curr_sled);
+                    tasks.spawn(self.check_instance(
+                        opctx,
+                        &curr_sled,
+                        &client,
+                        sled_instance,
+                    ));
+
+                    for (sled, sled_instance) in batch {
```

Comment:
I'm a bit confused why we have a `batch.next()` call _in addition_ to the loop here. Could these be collapsed into one loop, like a `while let` loop? A note about why it's structured this way may also be sufficient!

Reply:
We take the first element from the iterator by calling `batch.next()` before the `for` loop so that we can construct the initial sled-agent client with the sled address from the first item. I thought this seemed neater than making it an `Option` and having to handle the `None` case on every iteration.

I can add a comment explaining that!